Document NFT-related API methods

DennisDawson opened the pull request:
TOK-57: Add Docs for NFT Methods (https://ripplelabs.atlassian.net/browse/TOK-57) account_nfts, nft_buy_offers, nft_sell_offers

DennisDawson committed:
Add account_nfts method doc.

Per TOK-57.

DennisDawson committed:
Add doc for the nft_buy_offers method.

Per TOK-57.

DennisDawson committed:
Add doc for nft_sell_offers method

Per TOK-57.

DennisDawson committed:
Add NFT Methods

Add NFT Methods to the public-rippled-methods page, per TOK-57.

DennisDawson committed:
Add NFTs disclaimer.

mDuo13 commented on content/references/rippled-api/public-rippled-methods/nft-methods/account_nfts.md:
````diff
@@ -0,0 +1,118 @@
+    "method": "account_nfts",
+    "account": "rN7n7otQDd6FczFgLdSqtcsAUxDkw6fzRH"
+}
+```
````
You've mixed up JSON-RPC and WebSocket syntax in the examples of all three methods.

WebSocket:

- Has an `id` field (✅)
- Uses `command` for the method name (✅)
- Has other parameters as top-level fields (❌)

JSON-RPC:

- Uses the `method` field for the method name (✅)
- Has parameters nested in an object in a length-1 list in the `params` field (❌)

mDuo13 commented on content/references/rippled-api/public-rippled-methods/nft-methods/account_nfts.md:
```diff
@@ -0,0 +1,118 @@
+
+<!-- MULTICODE_BLOCK_END -->
+
+[Try it! >](websocket-api-tool.html#account_nfts)
```
These try-it links won't work automatically. The [apitool-methods-ws.js file](https://github.com/XRPLF/xrpl-dev-portal/blob/master/assets/js/apitool-methods-ws.js) needs to have entries for them.

Furthermore, the WebSocket Tool connects to Mainnet by default, where these methods won't work for a while. 

I put together [a draft commit](https://github.com/XRPLF/xrpl-dev-portal/commit/f63a23ea88c963262d34ff3cf2daf07f34812495) that extends the WS tool to accommodate these methods. It also adds examples for the NFT methods. If you pull that commit into this PR, then you can make the "Try it!" links work by adding a server query parameter that specifies NFT-Devnet, for example:

```suggestion
[Try it! >](websocket-api-tool.html?server=wss%3A%2F%2Fxls20-sandbox.rippletest.net%3A51233%2F#account_nfts)
```

mDuo13 commented on content/references/rippled-api/public-rippled-methods/nft-methods/account_nfts.md:
```diff
@@ -0,0 +1,118 @@
+| Field                 | Type                       | Description             |
+|:----------------------|:---------------------------|:------------------------|
+| `account`             | String                     | The unique identifier of an account, typically the account's [Address][]. The request returns a list of NFTs owned by this account. |
+|
```
Extraneous `|` here? Should be a blank line.

mDuo13 replied:
Also, there are more fields than you have documented here. Namely, `ledger_index`, `ledger_hash`, and `marker`, and `limit`, which work the same as in [account_objects](https://xrpl.org/account_objects.html).

mDuo13 commented on content/references/rippled-api/public-rippled-methods/nft-methods/account_nfts.md:
```diff
@@ -0,0 +1,118 @@
+      "validated": false
+    }
+  }
+}
```
There are a few syntax errors in this JSON (for example `Array(1)` isn't supposed to be there). Not sure if this output is from some tool, but we should fix it to be valid.

mDuo13 commented on content/references/rippled-api/public-rippled-methods/nft-methods/nft_buy_offers.md:
```diff
@@ -0,0 +1,111 @@
+labels:
+  - Non-fungible Tokens, NFTs, NFTokens
+---
+# account_nfts
```
Looks like you forgot to change the page title.

```suggestion
# nft_buy_offers
```

mDuo13 commented on content/references/rippled-api/public-rippled-methods/nft-methods/nft_sell_offers.md:
```diff
@@ -0,0 +1,112 @@
+labels:
+  - Non-fungible Tokens, NFTs, NFTokens
+---
+# account_nfts
```
```suggestion
# nft_sell_offers
```

mDuo13 commented on content/references/rippled-api/public-rippled-methods/nft-methods/account_nfts.md:
```diff
@@ -0,0 +1,118 @@
+---
+html: account_nfts.html
+parent: nft-methods.html
```
You need to create an entry in the `dactyl-config.yml` for the `nft-methods.html` landing page as well as for each of the 3 API method files, or they won't get built as part of the site. (That's where the link checker errors are coming from.)

Here's what the new entries should look like:

```yaml

        # TODO: translate blurb
    -   name: NFT Methods
        html: nft-methods.html
        parent: public-rippled-methods.html
        blurb: Use these methods to retrieve information about NFToken (non-fungible token) objects.
        template: pagetype-category.html.jinja
        targets:
            - en
            - ja

        # TODO: translate
    -   md: references/rippled-api/public-rippled-methods/nft-methods/account_nfts.md
        targets:
            - en
            - ja

        # TODO: translate
    -   md: references/rippled-api/public-rippled-methods/nft-methods/nft_buy_offers.md
        targets:
            - en
            - ja

        # TODO: translate
    -   md: references/rippled-api/public-rippled-methods/nft-methods/nft_sell_offers.md
        targets:
            - en
            - ja
```

That's assuming you go ahead with making the "NFT Methods" a category; in that case, this block should go in the big pages array where that category is, e.g. after the other related methods and before the `admin-rippled-methods.md` page.

If you like my suggestion below of working the methods into the existing categories, you can omit the first stanza, which defines a landing page for the section. In that case, you'd also want to update the frontmatter for each method to specify the correct category page as its `parent` attribute.

mDuo13 commented on content/references/rippled-api/public-rippled-methods/nft-methods/account_nfts.md:
```diff
@@ -0,0 +1,118 @@
+## Possible Errors
+
+* Any of the [universal error types][].
+* `rpcACT_NOT_FOUND` - The account string is not found on the ledger.
```
This is not the full list of possible errors; also, `rpcACT_NOT_FOUND` is the name in the source code but the value returned by the API is actually `actNotFound`.

There are some other possible errors not mentioned here; I have a strong suspicion that the full list for this method is exactly the same as the ones for account_objects, so you can probably copy-paste that one.

mDuo13 commented on content/references/rippled-api/public-rippled-methods/nft-methods/nft_buy_offers.md:
```diff
@@ -0,0 +1,111 @@
+---
+# account_nfts
+[[Source]](https://github.com/ripple/rippled/blob/xls20/src/ripple/rpc/handlers/NFTOffers.cpp "Source")
+{% include '_snippets/nfts-disclaimer.md' %}
```
I suggest moving the disclaimer below the one-line summary of the method (for all 3 files).

mDuo13 commented on content/references/rippled-api/public-rippled-methods/nft-methods/nft_buy_offers.md:
````diff
@@ -0,0 +1,111 @@
+    "tokenid": "000800001FD350FC4A6133360827DE142E60E7522E4F4DF5409177A90000000E"
+  },
+  "type": "response"
+}```
````
The ``` to end a code block should be on its own line.

mDuo13 commented on content/references/rippled-api/public-rippled-methods/nft-methods/nft_buy_offers.md:
```diff
@@ -0,0 +1,111 @@
+| `Field`        | Type             | Description                              |
+|:---------------|:-----------------|:-----------------------------------------|
+| `tokenid` | String | The NFToken associated with the sell offers |
+| `offers` | Array | The list of sell offers for the NFToken |
```
Please provide a definition table explaining the fields in each sell offer object. See [account_offers](https://xrpl.org/account_offers.html#response-format) for a comparable example.

mDuo13 commented on content/references/rippled-api/public-rippled-methods/nft-methods/nft_sell_offers.md:
```diff
@@ -0,0 +1,112 @@
+blurb: Get a list of all sell offers for a NFToken.
+labels:
+  - Non-fungible Tokens, NFTs, NFTokens
+---
```
Add `status: not_enabled` to the frontmatter (for all 3 methods) so the page gets the warning flask in the sidebar.

mDuo13 commented on content/references/rippled-api/public-rippled-methods/nft-methods/nft_buy_offers.md:
```diff
@@ -0,0 +1,111 @@
+
+<!-- MULTICODE_BLOCK_END -->
+
+[Try it! >](websocket-api-tool.html#nft_sell_offers)
```
Copy-paste error: this is the buy page, should link to the buy method.

```suggestion
[Try it! >](websocket-api-tool.html#nft_buy_offers)
```

mDuo13 commented on content/references/rippled-api/public-rippled-methods/nft-methods/nft_buy_offers.md:
```diff
@@ -0,0 +1,111 @@
+   "id": 1,
+   "command": "nft_buy_offers",
+   "params": [{
+   		"tokenid": "00090000D0B007439B080E9B05BF62403911301A7B1F0CFAA048C0A200000007"
```
This line mixes tab and space characters. There are several other examples in the files as well. Please don't use tab characters.

mDuo13 commented on content/references/rippled-api/public-rippled-methods/nft-methods/account_nfts.md:
```diff
@@ -0,0 +1,118 @@
+[[Source]](https://github.com/ripple/rippled/blob/xls20/src/ripple/rpc/handlers/AccountObjects.cpp "Source")
+{% include '_snippets/nfts-disclaimer.md' %}
+
+The `account_nfts` method returns a list of `NFToken` objects for the specified account.
```
As a best practice, the summary sentence should link back to a more general explanation for import keywords like "NFToken objects" and "account" in case people find their way to the API method without the relevant contextual information.

Many of the old API method pages _don't_ do this, which is a known issue (#1024), but we can at least not continue the problem in new pages.

mDuo13 commented on content/references/rippled-api/public-rippled-methods/public-rippled-methods.md:
```diff
@@ -97,6 +97,14 @@ Use these methods to perform convenient tasks, such as ping and random number ge
 * **[`ping`](ping.html)** - Confirm connectivity with the server.
 * **[`random`](random.html)** - Generate a random number.
 
+## [NFT Methods](nft-methods.html)
```
This is a reasonable way of categorizing the methods, but I would put it above the "Utility methods" and "Server Info Methods" because these are closer in purpose to the methods above those. Alternatively, you could add the 3 methods to existing categories: `account_nfts` would go with the other "Account Methods" (which makes sense, given that it starts with `account_` and functions similarly to those); and the buy/sell methods could go in the "Path and Order Book Methods" category, since they provide the same function for NFTs as the `book_offers` method does for fungible tokens.

mDuo13 commented on content/references/rippled-api/public-rippled-methods/nft-methods/account_nfts.md:
````diff
@@ -0,0 +1,118 @@
+}
+```
+
+<!-- MULTICODE_BLOCK_END -->
````
If there is a commandline format for these methods, please provide an example of that as well. If there is not a commandline method example, please add a note explaining that, such as the one on the [noripple_check page](https://xrpl.org/noripple_check.html).

mDuo13 committed:
WebSocket Tool: NFT methods & related support

- Add example syntax for the 3 NFT methods.
- Add the ability to specify the network without the rest of a
  permalink. NFT method "try it!" links can use this to have the tool
  use the NFT-Devnet by default.
- Add the ability to mark methods as not enabled on the production
  network.

mDuo13 commented on content/references/rippled-api/public-rippled-methods/nft-methods/nft_buy_offers.md:
```diff
@@ -0,0 +1,111 @@
+| Field                 | Type                       | Description             |
+|:----------------------|:---------------------------|:------------------------|
+| `tokenid`             | String                     | The unique identifier of a NFToken object. |
+|
```
There are other possible parameters for the request method; from my quick testing, `ledger_index` and `ledger_hash` are both possible. The [source code comments](https://github.com/ripple/rippled/blob/161bd6e9ad3fe25c0c6c5718835586124b9c1e2c/src/ripple/rpc/handlers/NFTOffers.cpp#L43-L44) suggest that `marker` and `limit` are also allowed, per the standard "markers and pagination" rules.

mDuo13 reviewed (changes requested):
I spotted several code errors as well as a few publishing issues that warrant your attention. Don't hesitate to reach out if you would like clarification or a more detailed explanation for any of these things.

mDuo13 committed:
WS tool: real NFT-Devnet address

mDuo13 committed:
Update NFT disclaimer

mDuo13 committed:
NFT API method fixes

mDuo13 committed:
NFT method fixes

mDuo13 commented:
Updated and rebased.
